mongorestore/mongorestore_test.go

```diff
@@ -69,6 +70,57 @@ func TestMongorestore(t *testing.T) {
 	}
 
 	Convey("With a test MongoRestore", t, func() {
+		// As specified in TOOLS-2363, the --db and --collection options
+		// are well-defined only for restoration of a single BSON file
+		Convey("and proper warning is issued if --db and --collection are not supported", func() {
+			// Hack to make log readable when testing
+
+			// TODO: Handle all logging in main so that ParseOptions
+			// and ParseAndValidateOptions can idiomatically return
+			// testable error values
+			var buffer bytes.Buffer
+
+			log.SetWriter(&buffer)
+			defer log.SetWriter(os.Stderr)
+
+			Convey("and no warning is issued in the well-defined case", func() {
+				// No error and nothing written in the log
+				args := []string{
+					"testdata/hashedIndexes.bson",
+					DBOption, "db1	",
+					CollectionOption, "coll1",
+				}
+
+				restore, err := getRestoreWithArgs(args...)
+
+				if err != nil {
+					t.Errorf("Cannot connect to test database: %v", err.Error())
+				}
+				err = restore.ParseAndValidateOptions()
+
+				So(err, ShouldBeNil)
+				So(buffer.String(), ShouldBeEmpty)
+			})
+
+			Convey("and a warning is issued in the deprecated case", func() {
+				// No error and some kind of warning message in the log
+				args := []string{
+					DBOption, "db1",
+					CollectionOption, "coll1",
+				}
+
+				restore, err := getRestoreWithArgs(args...)
+
+				if err != nil {
+					t.Errorf("Cannot connect to database: %v", err.Error())
+				}
+				err = restore.ParseAndValidateOptions()
+
+				So(err, ShouldBeNil)
+				So(buffer.String(), ShouldNotBeEmpty)
```

Comment:
I know it is hacky and not ideal, but should we do more than check that the output is non-empty? It looks like we do some direct error message matching in other tests. Again, that's not ideal but is relevant. See my comment above about the output of this error message. Without a direct test, you didn't catch that it is oddly formatted.

Reply:
The other tests are of the form `So(err.Error(), ShouldStartWith, "indextest.test collection: error...")`. I don't like this because you need one string literal for the message itself in core logic and then another for the message prefix in the test. This essentially duplicates the same invariant, which would make refactoring rather tedious in the future (you have to move things around in both places).

I think I can just write the warning as a global `const string` or some such, and then use it to directly check for equality. That way, we can maintain correctness of the code without depending on duplicated invariant literals like `TestDeprecatedIndexOptions` does.